### PR TITLE
Add Thermal Zones

### LINKED ISSUE #326: Temperature Monitor
**Is your feature request related to a problem? Please describe.**
I host my instance on a headless server, it's not very convenient to check my server temperature via SSH.

**Describe the solution you'd like**
Read thermal zone temperatures in ``/sys/class/thermal/thermal_zone*/temp`` and display temperature information in ServerInfo.

**Describe alternatives you've considered**
Nothing special.

**Additional context**
Currently WIP.
Screenshot_20210828_084023

### PR DESCRIPTION
**Pull request by wonderfulShrineMaidenOfParadise**
Fix #326 

## Preview

Screenshot_20210828_084023

Signed-off-by: wonderfulShrineMaidenOfParadise <wonderfulShrineMaidenOfParadise@users.noreply.github.com>

### CODE REVIEW

**Comment by kesselb** [comment]
Hey :wave: 

Thanks for your contribution. I'm not really a fan of this change. On most installation's it will not work anyway.

**Comment by wonderfulShrineMaidenOfParadise** [comment]
Okay. Maybe this could be improved?

**Comment by come-nc** [comment]
> Thanks for your contribution. I'm not really a fan of this change. On most installation's it will not work anyway.

Not work in what way? To be merge it should hide itself if no temperature can be measured.
@wonderfulShrineMaidenOfParadise any chance you can rebase on master?
Did you test the behavior when there is no thermal zone or when it returns nothing or negative value or such?

**Comment by kesselb** [comment]
> Not work in what way?

Afaik extra work is required to make the sensor information about temperature available. If we merge such a feature it's expected that it works and is tested. I don't see the benefit to make those information available via serverinfo. There are enough tools out there to setup a proper monitoring. Yet if you like this feature go ahead ;)

> any chance you can rebase on master?

In addition I would suggest the following changes. I reworked some of the information reading logic a while ago. I find it easier (and more reliable) to parse the information via php instead of piping some commands via shell_exec.

For example: `shell_exec('cat ' . $thermal_zone . '/type | tr -d \'\n\'');` is hard to read. Also cat and shell_exec is suspicious. I would try to use file_get_contents here. That should also work. See getMemory, getCpuName and getDiskInfo as example. Instead of shell_exec please use the readContent wrapper (to make testing possible).

**Comment by come-nc** [comment]
This is true about `shell_exec('cat`, we have a few instances of that that we should replace with direct file reading already.

**Comment by wonderfulShrineMaidenOfParadise** [comment]
- Rebased to solve conflicts
- Trying file_get_contents()
- If `/sys/class/thermal/thermal_zone*` doesn't exist the temperature section will be empty.

File: lib/OperatingSystems/DefaultOs.php [hunk 1]
```diff
@@ -179,6 +179,9 @@ public function getNetworkInterfaces(): array {
 		return $result;
 	}
 
+	/**
+	 * @return array
+	 */
```
Comment (kesselb) [inline lib/OperatingSystems/DefaultOs.php:184]:
```suggestion
```

Please remove the docblock. IOperatingSystem already contains the correct type hint.

**Comment by kesselb** [comment]
Thank you @wonderfulShrineMaidenOfParadise :+1:

**Comment by wonderfulShrineMaidenOfParadise** [comment]
- Use readContent()
- Fix coding styles

**Comment by kesselb** [comment]
Hey @wonderfulShrineMaidenOfParadise,

please revert the change to l10n/ja.json. Translation files are updated daily from transifex. All manual changes here are lost.

File: lib/OperatingSystems/DefaultOs.php [hunk 2]
```diff
@@ -215,6 +215,24 @@ public function getDiskInfo(): array {
 		return $data;
 	}
 
+	public function getThermalZones(): array {
+		$thermal_zones = glob('/sys/class/thermal/thermal_zone*');
```
Comment (PhrozenByte) [inline lib/OperatingSystems/DefaultOs.php:219]: `glob()` might return `false`, causing PHP errors in the iteration below

Comment (come-nc) [reply lib/OperatingSystems/DefaultOs.php:219]:
```suggestion
		$thermal_zones = glob('/sys/class/thermal/thermal_zone*') ?: [];
```
Same fix as the other instance by @PhrozenByte should be enough

File: templates/settings-admin.php [hunk 3]
```diff
@@ -69,6 +67,27 @@ function FormatMegabytes(int $byte): string {
...
+				</h2>
+				<div class="table-wrapper">
+					<table class="server-infos-table">
+						<thead>
+						</thead>
+						<tbody>
+						<?php foreach ($_['thermalzones'] as $thermalzone): ?>
+						<tr>
+							<td><?php p($thermalzone['type']) ?>:</td>
+							<td><span class="info" id="<?php p($thermalzone['type']) ?>_temp"></span>°C</td>
```
Comment (come-nc) [inline templates/settings-admin.php:84]: Are we 100% sure that there can be no duplicate type?
Should we maybe use types as key in the array returned by getThermalZones to make that requirement clear?

**Review by come-nc — APPROVED** [review]
Two small concerns but otherwise good to go.

File: templates/settings-admin.php (hunk 3 above)
Comment (wonderfulShrineMaidenOfParadise) [reply templates/settings-admin.php:84]: Duplicate type? Could you describe how it looks like?
There should be only one file `type` for each thermalzone, `/sys/class/thermal/thermal_zone*/temp`.

File: lib/OperatingSystems/DefaultOs.php (hunk 2 above)
Comment (wonderfulShrineMaidenOfParadise) [reply lib/OperatingSystems/DefaultOs.php]: @come-nc Applied. Thank you for fixing it.

File: templates/settings-admin.php (hunk 3 above)
Comment (wonderfulShrineMaidenOfParadise) [reply templates/settings-admin.php:84]:
```
$ cat /sys/class/thermal/thermal_zone*/{type,temp}
acpitz
acpitz
pch_haswell
INT3400 Thermal
x86_pkg_temp
52800
52800
49500
20000
52000
```

Indeed I see this on my laptop. Not sure if those 2 `acpitz` are the same thing, though.

Comment (kesselb) [reply templates/settings-admin.php:84]: To update the data in the frontend properly on an xhr update we need some to identify the right sensor. Please add add a new property 'zone' to the $tzone array and set md5($thermal_zone) as value. The hashed path should be stable enough. 

And please change snake_case (e.g. $thermal_zones) to camelCase ($thermalZones).

Thanks!

File: lib/OperatingSystems/DefaultOs.php [hunk 4]
```diff
@@ -215,6 +215,25 @@ public function getDiskInfo(): array {
...
 	}
 
+	public function getThermalZones(): array {
+		$thermalZones = glob('/sys/class/thermal/thermal_zone*') ?: [];
+		$result = [];
+
+		foreach ($thermalZones as $thermalZone) {
+			$tzone = [];
+			try {
+				$tzone['zone'] = md5($thermalZone);
```
Comment (come-nc) [inline lib/OperatingSystems/DefaultOs.php:225]: I would name it `id` (or `hash` or `sum`).

File: lib/OperatingSystems/IOperatingSystem.php [hunk 5]
```diff
@@ -91,4 +91,18 @@ public function getTime(): string;
 	 * Get the total number of seconds the system has been up or -1 on failure.
 	 */
 	public function getUptime(): int;
+
+	/**
+	 * Get info about available thermal zones.
+	 *
+	 * [
+	 *       [
+	 *             'type' => string,
```
Comment (come-nc) [inline lib/OperatingSystems/IOperatingSystem.php:100]: Add sum in here as well

Comment (wonderfulShrineMaidenOfParadise) [reply lib/OperatingSystems/IOperatingSystem.php:101]: Pushed another one.

File: templates/settings-admin.php (hunk 3 above)
Comment (kesselb) [reply templates/settings-admin.php:84]: Thank you very much!

The hopefully last step is to change

`<td><span class="info" id="<?php p($thermalzone['type']) ?>_temp"></span>°C</td>`

to 

`<td><span class="info" id="<?php p($thermalzone['hash']) ?>"></span>°C</td>`

(to use hash as unique identifier for the html elment)

and update `script.js` to also use the hash

`document.getElementById(data.thermalzones[i]['type'] + "_temp").innerHTML = data.thermalzones[i]['temp'];`

to 

`document.getElementById(data.thermalzones[i]['hash']).innerHTML = data.thermalzones[i]['temp'];`

(note: I don't have thermal sensors in my /sys so I cannot test myself)

Comment (wonderfulShrineMaidenOfParadise) [reply templates/settings-admin.php]: Done. It still looks good for my sensors. Please do let me know if there is something can be improved.

Comment (kesselb) [reply templates/settings-admin.php]: Thank you very much!

**Comment by come-nc** [comment]
:rocket:

**Comment by come-nc** [comment]
@wonderfulShrineMaidenOfParadise Thank you for your work and patience :heart:

**Comment by ostasevych** [comment]
I have an old hp proliant n40l and it doesn't have a class thermal_zone0 at all, but
`/sys/devices/pci0000:00/0000:00:18.3/hwmon/hwmon3/temp1_crit
/sys/devices/pci0000:00/0000:00:18.3/hwmon/hwmon3/temp1_crit_hyst
/sys/devices/pci0000:00/0000:00:18.3/hwmon/hwmon3/temp1_input
/sys/devices/pci0000:00/0000:00:18.3/hwmon/hwmon3/temp1_max
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp5_crit_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp2_type
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp5_source_sel
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp2_crit
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp2_max
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp1_type
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp3_source_sel
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp1_crit
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp1_crit_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp1_source_sel
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp6_source_sel
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp5_max_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp1_alarm
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp5_type
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp5_alarm
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp1_input
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp1_max_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp5_crit
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp5_input
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp2_crit_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp2_alarm
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp2_source_sel
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp2_max_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp1_max
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp5_max
/sys/devices/pci0000:00/0000:00:14.0/i2c-1/1-002f/temp2_input
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0019/hwmon/hwmon1/temp1_crit
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0019/hwmon/hwmon1/temp1_crit_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0019/hwmon/hwmon1/temp1_min_alarm
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0019/hwmon/hwmon1/temp1_input
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0019/hwmon/hwmon1/temp1_max_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0019/hwmon/hwmon1/temp1_crit_alarm
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0019/hwmon/hwmon1/temp1_min
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0019/hwmon/hwmon1/temp1_max_alarm
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0019/hwmon/hwmon1/temp1_max
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0018/hwmon/hwmon0/temp1_crit
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0018/hwmon/hwmon0/temp1_crit_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0018/hwmon/hwmon0/temp1_min_alarm
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0018/hwmon/hwmon0/temp1_input
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0018/hwmon/hwmon0/temp1_max_hyst
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0018/hwmon/hwmon0/temp1_crit_alarm
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0018/hwmon/hwmon0/temp1_min
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0018/hwmon/hwmon0/temp1_max_alarm
/sys/devices/pci0000:00/0000:00:14.0/i2c-0/0-0018/hwmon/hwmon0/temp1_max`

Is it possible to use these data as well to show the temperature of the system hardware?

**Comment by wonderfulShrineMaidenOfParadise** [comment]
@ostasevych this PR should work for PC and phones. what do you see in `/sys/class/thermal/`?

**Comment by ostasevych** [comment]
> @ostasevych this PR should work for PC and phones. what do you see in `/sys/class/thermal/`?

Hi. Nothing at all.

**Comment by wonderfulShrineMaidenOfParadise** [comment]
Currently it can only parse subsystem `thermal`. Maybe you can try to implement your solution.
Unfortunately I don't have any plan on serverinfo for now. Grafana would be better for monitoring.

**Comment by ostasevych** [comment]
> Currently it can only parse subsystem `thermal`. Maybe you can try to implement your solution. Unfortunately I don't have any plan on serverinfo for now. Grafana would be better for monitoring.

Hi! Thanks, of course... In case the server is behind the VPN it might be difficult to use any external tool.

**Pull request merged**